src/lib.rs

```diff
@@ -231,78 +266,73 @@ impl BodyGenerator for IntoOwnedGen {
-                    })
-                    .collect::<Vec<_>>();
-                quote! { { #(#fields),* } }
-            },
-            syn::VariantData::Tuple(ref body) => {
-                let fields = body.iter()
-                    .enumerate()
-                    .map(|(index, field)| {
-                        let index = syn::Ident::from(index);
-                        let index = quote! { self.#index };
-                        FieldKind::resolve(field).borrow_or_clone(&index)
-                    })
-                    .collect::<Vec<_>>();
-                quote! { ( #(#fields),* ) }
-            },
-            syn::VariantData::Unit => {
-                quote! {}
-            }
+    fn visit_enum_data(
+        &self,
+        ident: proc_macro2::TokenStream,
+        variant: &syn::Variant,
+    ) -> proc_macro2::TokenStream {
+        if variant.fields.is_empty() {
+            return quote!(#ident => #ident);
         }
-    }
 
-    fn visit_enum_data(&self, ident: quote::Tokens, data: &syn::VariantData) -> quote::Tokens {
-        match *data {
-            syn::VariantData::Struct(ref body) => {
-                let idents = body.iter()
-                    .map(|field| field.ident.as_ref().unwrap())
-                    .collect::<Vec<_>>();
-                let cloned = body.iter()
-                    .map(|field| {
-                        let ref ident = field.ident.as_ref().unwrap();
-                        let ident = quote! { #ident };
-                        let code = FieldKind::resolve(field).borrow_or_clone(&ident);
-                        quote! { #ident: #code }
-                    })
-                    .collect::<Vec<_>>();
-                quote! { #ident { #(ref #idents),* } => #ident { #(#cloned),* } }
-            },
-            syn::VariantData::Tuple(ref body) => {
-                let idents = (0..body.len())
-                    .map(|index| syn::Ident::from(format!("x{}", index)))
-                    .collect::<Vec<_>>();
-                let cloned = idents.iter().zip(body.iter())
-                    .map(|(ident, field)| {
-                        let ident = quote! { #ident };
-                        FieldKind::resolve(field).borrow_or_clone(&ident)
-                    })
-                    .collect::<Vec<_>>();
-                quote! { #ident ( #(ref #idents),* ) => #ident ( #(#cloned),* ) }
-            },
-            syn::VariantData::Unit => {
-                quote! { #ident => #ident }
-            },
+        let fields_are_named = variant.fields.iter().any(|field| field.ident.is_some());
+        if fields_are_named {
+            let idents = variant
+                .fields
+                .iter()
+                .map(|field| field.ident.as_ref().expect("this fields has no ident (5)"));
+            let cloned = variant.fields.iter().map(|field| {
+                let ident = field.ident.as_ref().expect("this fields has no ident (6)");
```

Comment:
This split into named and unnamed fields seems rather strange. Did you pick up any references on this issue? Feels like there has to be a better way (tm) to handle it.

Not suggesting that it should be fixed on this PR, just asking since you've now researched this topic quite heavily.

Reply:
I was not able to give the issue the amount of attention I would have liked, it was over the xmas holidays after all, so there is still a lot I don't know about proc_macros and best practices. This can probably be done better. However it also sorta resembles the match on `VariantData::Struct` and `VariantData::Tuple` that was there [before](https://github.com/koivunej/derive-into-owned/blob/master/src/lib.rs#L189), so semantically it does the same thing.

🤷‍♂️

Reply:
Yeah I can't remember running into any ideas on this, but maybe we'll find it later on.